rfc/indexing-design.md

```diff
@@ -0,0 +1,228 @@
+# W3 Indexing Design
+> Replaces [W3 IPNI Indexing RFC](https://github.com/w3s-project/RFC/blob/main/rfc/ipni-w3c.md)
+
+## Authors
+
+- [gammazero]
+- [gozala]
+
+## Goals
+
+A user can use a multihash, that identifies data stored on w3s, to lookup the all information necessary to retrieve and delegate access to that data.
+
+Location commitments that describe where content is stored can change over time as w3s moves data to a more favorable location, re-aggregates data storage, or replicates stored data. Location commitments are updated without having to index all multihashes.
+
+A query API will allow a user to specify indexed content to lookup and how to filter results. The API retrieves previously cached responses or composes new IPNI queries to get new responses to filter and cache.
+
+## Overview
+
+A user stores content as opaque blobs containing series of blocks addressed by multihashes. These blocks are collected into one or more CAR file (shards) and the CAR file is stored at a location designated for a particular user's data.
+
+The user creates a blob-index ([sharded-dag-index](https://github.com/w3s-project/specs/blob/main/w3-index.md#sharded-dag-index-example)) that contains the multihash of each shard of blob data, and an array of tuples of (multihash, offset, length) for all the blocks in each shard. The user may choose which blocks to include in the blob-index if they do not want all blocks to be indexed. For example, they may want only the DAG root to be indexed.
+
+The blob-index is stored in the user's space along with the CAR file(s) containing the blob data. This is done in the [w3 add-blob operation](https://github.com/w3s-project/specs/blob/main/w3-blob.md#add-blob). The w3s service creates [location commitments](https://github.com/w3s-project/specs/blob/main/w3-blob.md#location-commitment) for the blob shards and for the blob-index. 
```

Comment:
One reason to keep them separate is because the index describes all shards, and shards may be stored in separate locations. If only the 3rd shard is needed then only the index and that shard need to be read.

It may be acceptable to encode the index as part of the first shard... but then how does the client know how much of the first shard to read? Would there be a fixed-length header?